### DISCUSSION TITLE
Help understanding pointers and more.

### DISCUSSION THREAD

**Discussion by ghost** [community/community#80835] [Programming Help] [answered]
### Body

```cpp
#include <iostream>
#include <limits.h>
using namespace std;
int main(){
    int *a = nullptr;
    a = new int;
    int* memoryLocation = (int*)a;
    *memoryLocation = 100101010;
    cout << memoryLocation; //0x5628f75caeb0
    
}
```
Also,

Where can I learn more low-level concepts?

### Guidelines

- [x] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by mesh-dell** [comment, marked as answer]
int *a = nullptr;: Declares a pointer a to an int and initializes it to nullptr, indicating it doesn't point to any allocated memory yet.
a = new int;: Allocates memory for an int on the heap and assigns its address to a. This memory is managed dynamically and needs to be released later to avoid memory leaks.
int* memoryLocation = (int*)a;: Creates another pointer memoryLocation and casts the type of a (already pointing to an int) to explicitly declare it as an int*.
*memoryLocation = 100101010;: Assigns the value 100101010 to the memory location pointed to by memoryLocation (which is the same memory as a).
cout << memoryLocation;: Prints the address of the allocated memory, in this case, "0x5628f75caeb0".
Learning low-level concepts:

Books:
"C Programming Language" by Brian Kernighan and Dennis Ritchie
"Computer Systems: A Programmer's Perspective" by Randal E. Bryant and David O'Hallaron
Online resources:
MIT OpenCourseware: 6.S191 Introduction to Computer Systems
Stanford CS106L Programming Methodology with Java
TutorialsPoint: Dynamic Memory Allocation in C++
Practice:
Experiment with dynamic memory allocation in C++ and other languages.
Solve coding challenges and exercises that focus on low-level concepts.
Remember, understanding memory management and dynamic allocation is crucial for efficient and safe programming. Take your time, explore the resources, and practice to solidify your understanding!
